crypto/threads_pthread.c

```diff
@@ -638,8 +716,34 @@ __owur int CRYPTO_THREAD_read_lock(CRYPTO_RWLOCK *lock)
 __owur int CRYPTO_THREAD_write_lock(CRYPTO_RWLOCK *lock)
 {
 # ifdef USE_RWLOCK
+#  ifdef REPORT_RWLOCK_CONTENTION
+    if (pthread_rwlock_trywrlock(lock)) {
+        int nptrs;
+        void *buffer[BT_BUF_SIZE];
+        char **strings;
+        OSSL_TIME start, end;
+
+        start = ossl_time_now();
+        if (!ossl_assert(pthread_rwlock_wrlock(lock) == 0))
+            return 0;
+        end = ossl_time_now();
+        if (contention_fp == NULL)
+            return 1;
+        nptrs = backtrace(buffer, BT_BUF_SIZE);
+        strings = backtrace_symbols(buffer, nptrs);
+        fprintf(contention_fp, "lock blocked on WRITE for %zu usec\n",
+                ossl_time2us(ossl_time_subtract(end, start)));
+        if (strings != NULL) {
+            for (int j = 0; j < nptrs; j++)
+                fprintf(contention_fp, "%s\n", strings[j]);
+            free(strings);
+        }
+        fprintf(contention_fp, "\n");
```

Comment:
It seems there should be an opportunity to have a common function to write this stuff to the log file - rather than having two copies of the same code